language/functional_tests/tests/testsuite/discovery/reconfiguration_via_address_rotation.mvir

```diff
@@ -0,0 +1,142 @@
+//! account: alice
+//! account: vivian, 1000000, 0, validator
+//! account: viola, 1000000, 0, validator
+
+//! new-transaction
+//! sender: alice
+import 0x0.ValidatorConfig;
+
+// register Alice as a validator candidate
+main() {
+    let config: ValidatorConfig.Config;
+
+    ValidatorConfig.register_candidate_validator(h"10", h"20", h"30", h"40", h"50", h"60");
+    return;
+}
+
+// check: EXECUTED
+
+//! new-transaction
+//! sender: alice
+import 0x0.LibraSystem;
+import 0x0.ValidatorConfig;
+
+// rotate alice's network address
+main() {
+    ValidatorConfig.rotate_validator_network_address(h"40");
+    return;
+}
+
+// check: events: []
+// check: EXECUTED
+
+//! new-transaction
+//! sender: vivian
+import 0x0.LibraSystem;
+
+// Run the block prologue. No reconfiguration should be triggered, since alice is
+// not a validator
+// TODO: trigger a BlockMetadata transaction here instead
+main() {
+    LibraSystem.block_prologue(1, h"aa", h"bb", {{vivian}});
+    return;
+}
+// not: ValidatorSetChangeEvent
+// not: DiscoverySetChangeEvent
+// check: EXECUTED
+
+//! new-transaction
+//! sender: vivian
+import 0x0.LibraSystem;
+import 0x0.ValidatorConfig;
+
+// rotate vivian's network address and then run the block prologue. Now,
+// reconfiguration should be triggered.
+main() {
+    let info: LibraSystem.DiscoveryInfo;
+
+    ValidatorConfig.rotate_validator_network_address(h"40");
+    // check that the discovery set contains Vivian's old address before reconfiguration
+    info = LibraSystem.get_ith_discovery_info(1);
+    assert(*LibraSystem.get_validator_network_address(&info) != h"40", 98);
+    LibraSystem.block_prologue(2, h"aa", h"bb", {{vivian}});
```

Comment:
In the near future, `block_prologue` will be private and we'll have to write tests like this by using a directive like `//! new-block.`. This refactoring will be less painful if each transaction that calls `block_prologue` is standalone (like your previous tx). However, I realize splitting this will be more annoying for you, so no worries if you don't want to take it on.